### DISCUSSION TITLE
On M/M/k/k optimization problem implementation

### DISCUSSION THREAD

**Discussion by Hadi2525** [cvxpy/cvxpy#1449] [Q&A]
Hello,

I would like to formulate a queueing network (M/M/k/k) optimization problem with the number of servers (k) as the variable in cvxpy. Since the objective function (i.e. blocking probability) is convex so I am optimistic to solve it in cvxpy, however, there is no atomic function available to help me with that.

I realize there is an example of the design of a queueing system on the [website](https://www.cvxpy.org/examples/derivatives/queuing_design.html) when the variable becomes the number of servers, then it becomes hard to solve. This problem is somewhat similar to the optimization problem (5) of [Mung Chiang's paper](https://stanford.edu/~boyd/papers/pdf/nonlin_opt_queue.pdf) with the difference here that arrival/service rates are constant and the number of servers is the optimization variables.

How is it possible to implement this problem in cvxpy?

**Comment by SteveDiamond** [comment]
@akshayka what do you think? @Hadi2525 I'm a bit confused, do either the formulation on the website or in Mung Chiang's paper relate to your problem? They both can be written in cvxpy fairly easily.

**Comment by Hadi2525** [comment]
@SteveDiamond, the problem I have is that my variable is the number of servers (i.e. integer `m` in eqn (5) of Mung Chiang's paper) while arrival/service rates are constant. If `m` was known, the formulation was not a big issue.

**Reply by SteveDiamond** [reply]
Is number of servers the only variable? Can you just do binary search on the number of servers?

**Reply by SteveDiamond** [reply]
To elaborate a bit, you might find ``length`` useful: https://www.cvxpy.org/examples/dqcp/minimum_length_least_squares.html

**Reply by Hadi2525** [reply]
@SteveDiamond, Thank you for the reply. Yes, the number of servers is the variable. I am not sure if the BS would be helpful.

The `length` problem is not similar to the obj. func. I am dealing with.

**Reply by angeris** [reply]
sorry, why would something like ternary search not work here? Isn't the function unimodal?

**Reply by Hadi2525** [reply]
Since the objective function is convex I want to formulate the problem as a convex optimization problem and solve it in cvxpy.

**Reply by angeris** [reply]
Ok, I guess I'm just not understanding what your formulation is here. What I assumed was that you had something like (5) except that \lambda and \mu are constant, in which case there is no optimization problem (for fixed m) because you can simply choose `C_j` to be any value between `C_{j, min}` and `p_j`. (If there is no such value, or the \mu, \lambda don't satisfy the constraints, then there is no solution.)

The optimal value of this problem is also fixed, since p_k is fixed, and now you just have to search through the parameter m, which can easily be done with ternary search. I'm also not sure how this problem is convex, since m is integer-valued?

**Reply by Hadi2525** [reply]
By relaxing the assumption that `m` is a real positive number, `p_k` is a convex function w.r.t. `m`. Obviously, the proof of convexity for `p_k` is not straightforward. This can be verified through plotting `p_k` for values of `m`.

The formulation is as in (5) of the paper, however, the only variable we have is `m` with some constraints. The constraints obviously limit the feasible space for `m`, then in the feasible space with a brute force search (the worst than binary or ternary search), the global optimum point can be found, right? So, then how can we formulate the problem as a convex problem in cvxpy?

**Reply by angeris** [reply]
In (5), m is not a variable, as it is fixed, so I'm not really sure what you mean by the latter paragraph?

On the other hand, we know the solution to what I think is the problem you're describing, directly. Given some fixed k, just make m as large as possible, so long as m is 'feasible.' I.e., so long as there exists a feasible point for (5) with this m. (This is optimal is because the denominator of p_k is strictly increasing in m.)

If I'm understanding the problem correctly, I think this should be your solution, but I'm not 100% sure I still quite get it.

Additionally, it's not clear how we can relax m to be positive real, rather than just an integer?

**Reply by Hadi2525** [reply]
Thank you @angeris. 

The simple version of the problem that I am solving is to consider `m` in (5) as a variable while lambda and mu are fixed.

If you look at large values of `m`, then the decrease in blocking probability is not significant (for some value of m onward the change is insignificant), and therefore the cost of installing excessive servers will exceed some installation constraints.

The original problem is this: we want to design a service provider with k servers composed of 3 server types (k = k1 + k2 + k3) each with a unique service rate (mu1, mu2, m3). The question is given a certain arrival and service rate, how many servers of each type is required for a minimum blocking probability such that we have an optimum number of each server already installed and we have limited funding available for the investment cost of k servers (k1 >= k1min, k2 >= k2min, k3 >= k3min | c1k1 + c2k2 +c3k3 <= C with C as our investment cost threshold).

I am looking at the obj func as an Erlang formula in a continuous domain. I hope that helps to understand the question. Let me know @angeris if you need more information.
